network/router_test.go

```diff
@@ -0,0 +1,365 @@
+package network
+
+import (
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/dedis/cothority/log"
+	"github.com/stretchr/testify/assert"
+)
+
+func NewTestRouterTCP(port int) (*Router, error) {
+	h, err := NewTestTCPHost(port)
+	if err != nil {
+		return nil, err
+	}
+	id := NewTestServerIdentity(h.addr)
+	return NewRouter(id, h), nil
+}
+
+func NewTestRouterLocal(port int) (*Router, error) {
+	h, err := NewTestLocalHost(port)
+	if err != nil {
+		return nil, err
+	}
+	id := NewTestServerIdentity(h.addr)
+	return NewRouter(id, h), nil
+}
+
+type routerFactory func(port int) (*Router, error)
+
+// Test if router fits the interface such as calling Run(), then Stop(),
+// should return
+func TestRouterTCP(t *testing.T) {
+	testRouter(t, NewTestRouterTCP)
+}
+func TestRouterLocal(t *testing.T) {
+	testRouter(t, NewTestRouterLocal)
+}
+
+func testRouter(t *testing.T, fac routerFactory) {
+	h, err := fac(2004)
+	if err != nil {
+		t.Fatal(err)
+	}
+	var stop = make(chan bool)
+	go func() {
+		stop <- true
+		h.Start()
+		stop <- true
+	}()
+	<-stop
+	// Time needed so the listener is up. Equivalent to "connecting ourself" as
+	// we had before.
+	time.Sleep(250 * time.Millisecond)
+	h.Stop()
+	select {
+	case <-stop:
+		return
+	case <-time.After(500 * time.Millisecond):
+		t.Fatal("TcpHost should have returned from Run() by now")
+	}
+}
+
+// Test the automatic connection upon request
+func TestRouterAutoConnectionTCP(t *testing.T) {
+	testRouterAutoConnection(t, NewTestRouterTCP)
+}
+func TestRouterAutoConnectionLocal(t *testing.T) {
+	testRouterAutoConnection(t, NewTestRouterLocal)
+}
+
+func testRouterAutoConnection(t *testing.T, fac routerFactory) {
+	//	log.TestOutput(true, 5)
+	fmt.Println("Before fac()")
+	h1, err := fac(2007)
+	if err != nil {
+		t.Fatal(err)
+	}
+	fmt.Println("after fac()")
+	err = h1.Send(&ServerIdentity{Address: NewLocalAddress("127.1.2.3:2890")}, &SimpleMessage{12})
+	if err == nil {
+		t.Fatal("Should not be able to send")
+	}
+	fmt.Println("after send() #1")
+	h2, err := fac(2008)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = h1.Send(h2.id, nil)
+	if err == nil {
```

Comment:
But it's not the same behavior. `log.ErrFatal` will fatal and `t.Fatal` calls `t.FailNow` which does:
`FailNow marks the function as having failed and stops its execution. Execution will continue at the next test or benchmark.`